specs/capella/beacon-chain.md

```diff
@@ -323,11 +328,18 @@ def process_withdrawals(state: BeaconState, payload: ExecutionPayload) -> None:
     for expected_withdrawal, withdrawal in zip(expected_withdrawals, payload.withdrawals):
         assert withdrawal == expected_withdrawal
         decrease_balance(state, withdrawal.validator_index, withdrawal.amount)
-    if len(expected_withdrawals) > 0:
-        latest_withdrawal = expected_withdrawals[-1]
-        state.next_withdrawal_index = WithdrawalIndex(latest_withdrawal.index + 1)
-        next_validator_index = ValidatorIndex((latest_withdrawal.validator_index + 1) % len(state.validators))
-        state.next_withdrawal_validator_index = next_validator_index
+
+    if len(expected_withdrawals) == 0:
```

Comment:
```suggestion
    if len(expected_withdrawals) < MAX_WITHRAWALS_PER_PAYLOAD:
```

Reply:
if there are more than 0 withdrawals, then the usual process takes over where you just go via the last validator you could get to

if we _always_ skip forward by the sweep bound then we would end up jumping around the validator set and lose the sequential nature of the scan

Reply:
If you didn't achieve the maximum then we know for a fact that no validator was withdrawable this block in that sweep. There's no need to rescan in the next slot just to include the possible one that may have become withrawable during this one. It's more effective to just add here and continue from where we scanned in this loop

Reply:
In other words, given your argument, why not just add 1 then when there are zero, instead of the full max bound? There may be validators that become withrawable in the next slot and adding one already keeps us alive

Reply:
> If you didn't achieve the maximum then we know for a fact that no validator was withdrawable this block in that sweep.

no?

let's start at validator index 0, sweep 256 with block max of 16. 

if you get 0 withdrawals, then no validator was withdrawable so we advance the state of the sweep by 256
if we get 16 withdrawals, then it seems most fair to just start at validator index 16 next time
if you get some other number of withdrawals, then I think you want to start just after the index of the last withdrawal you got...

the only special case is when you have 0 withdrawals which this PR currently handles

I would not want to just add 1 to the validator index as that seems that we would duplicate inspections of validators per epoch (keep in mind that withdrawable status only changes at epoch boundaries, unless there is a deposit or if you are slashed which I'll argue are infrequent enough to not bias the 'fairness' of this choice of scan significantly)

Reply:
> if you get some other number of withdrawals, then I think you want to start just after the index of the last withdrawal you got...
> 
I disagree with this, if you get less than 16 it means that you scanned 256 and found no withdrawable validators and you are making the point below that it's a good assumption that they will not be withdrawable the next slot, so I want to start next sweep at 256 more, and not the last index. 

> I would not want to just add 1 to the validator index as that seems that we would duplicate inspections of validators per epoch (keep in mind that withdrawable status only changes at epoch boundaries, unless there is a deposit or if you are slashed which I'll argue are infrequent enough to not bias the 'fairness' of this choice of scan significantly)

Reply:
ok I see your point and I'll update it, thanks for explaining!